### PR TITLE
[Uploading models guide] Update docs

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR improves  the "uploading models" guide a bit more, as the `huggingface_hub` library recently has improved quite a bit.

### CODE REVIEW

File: docs/hub/models-uploading.md [hunk 1]
```diff
@@ -7,10 +7,14 @@ You can link repositories with an individual user, such as [osanseviero/fashion_
...
+- In case your model is a custom PyTorch model, there are 2 ways:
+    - using the [custom code](https://huggingface.co/docs/transformers/custom_models) feature on the hub, which makes your model usable in the Transformers library
+    - leveraging the [huggingface_hub](#upload-a-pytorch-model-using-huggingface_hub) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module`, just like models in the Transformers, Diffusers and Timm libraries.
 - In addition to programmatic uploads, you can always use the [web interface](#using-the-web-interface).
 
-Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model.
+Do note that only the first 2 ways will make sure that download metrics work for your model.
+
+Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model and make it easier discoverable.
```
Comment (Wauplin) [inline docs/hub/models-uploading.md:17]:
```suggestion
Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model and make it more discoverable.
```

File: docs/hub/models-uploading.md [hunk 2]
```diff
@@ -102,29 +106,29 @@ from huggingface_hub import PyTorchModelHubMixin
...
 # push to the hub
-model.push_to_hub("my-awesome-model", config=config)
+model.push_to_hub("my-awesome-model")
 
 # reload
 model = MyModel.from_pretrained("username/my-awesome-model")
 ```
 
-As can be seen, the only thing required is to define all hyperparameters regarding the model architecture (such as hidden size, number of classes, dropout probability, etc.) in a Python dictionary often called the `config`. Next, you can define a class which takes the `config` as keyword argument in its init.
+As can be seen, the only thing required is to inherit from `PyTorchModelHubMixin`. All attributes will be automatically serialized as a `config.json`.
```
Comment (Wauplin) [inline docs/hub/models-uploading.md:131]:
```suggestion
As can be seen, the only thing required is to inherit from `PyTorchModelHubMixin`. All attributes will be automatically serialized in a `config.json` file.
```

File: docs/hub/models-uploading.md [hunk 3]
```diff
@@ -7,10 +7,14 @@ You can link repositories with an individual user, such as [osanseviero/fashion_
...
 
 - In case your model comes from a library that has [built-in support](#upload-from-a-library-with-built-in-support), one can use the existing methods.
-- In case your model is a custom PyTorch model, the recommended way is to leverage the [huggingface_hub](#using-the-huggingface_hub-client-library) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to your models, just like models in the Transformers, Diffusers and Timm libraries.
+- In case your model is a custom PyTorch model, there are 2 ways:
+    - using the [custom code](https://huggingface.co/docs/transformers/custom_models) feature on the hub, which makes your model usable in the Transformers library
+    - leveraging the [huggingface_hub](#upload-a-pytorch-model-using-huggingface_hub) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module`, just like models in the Transformers, Diffusers and Timm libraries.
```
Comment (Wauplin) [inline docs/hub/models-uploading.md:12]:
```suggestion
- In case your model is a custom PyTorch model, the two recommended ways to integrate with the ecosystem is to either:
    - use the `transformers`'s [custom code](https://huggingface.co/docs/transformers/custom_models) feature, which makes your model directly usable in the Transformers library
    - or leverage the [`huggingface_hub`'s `PyTorchModelHubMixin` class](#upload-a-pytorch-model-using-huggingface_hub) as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module` class, just like models in the Transformers, Diffusers and Timm libraries.
```

File: docs/hub/models-uploading.md [hunk 4]
```diff
@@ -7,10 +7,14 @@ You can link repositories with an individual user, such as [osanseviero/fashion_
 There are several ways to upload models to the Hub, described below.
 
 - In case your model comes from a library that has [built-in support](#upload-from-a-library-with-built-in-support), one can use the existing methods.
-- In case your model is a custom PyTorch model, the recommended way is to leverage the [huggingface_hub](#using-the-huggingface_hub-client-library) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to your models, just like models in the Transformers, Diffusers and Timm libraries.
+- In case your model is a custom PyTorch model, the two recommended ways to integrate with the ecosystem is to either:
+    - use the `transformers`'s [custom code](https://huggingface.co/docs/transformers/custom_models) feature, which makes your model directly usable in the Transformers library
```
Comment (julien-c) [inline docs/hub/models-uploading.md:11]: this is a bit too transformers specific at this stage no?

Comment (julien-c) [reply docs/hub/models-uploading.md:11]: the rest LGTM from a quick read

File: docs/hub/models-uploading.md [hunk 5]
```diff
@@ -7,10 +7,14 @@ You can link repositories with an individual user, such as [osanseviero/fashion_
 There are several ways to upload models to the Hub, described below.
 
 - In case your model comes from a library that has [built-in support](#upload-from-a-library-with-built-in-support), one can use the existing methods.
-- In case your model is a custom PyTorch model, the recommended way is to leverage the [huggingface_hub](#using-the-huggingface_hub-client-library) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to your models, just like models in the Transformers, Diffusers and Timm libraries.
+- In case your model is a custom PyTorch model, the two recommended ways to integrate with the ecosystem is to either:
```
Comment (osanseviero) [inline docs/hub/models-uploading.md:10]: WDYT about adding a bolded message in the first paragraph clarifying that repos don't need to be `transformers`-based

File: docs/hub/models-uploading.md (hunk 4 above)
Comment (osanseviero) [reply docs/hub/models-uploading.md:11]: Agree, and I see this more of a `model comes from a library that has ...` class. E.g. we also have remote code with `diffusers`.

File: docs/hub/models-uploading.md [hunk 6]
```diff
@@ -7,10 +7,14 @@ You can link repositories with an individual user, such as [osanseviero/fashion_
...
+    - use the `transformers`'s [custom code](https://huggingface.co/docs/transformers/custom_models) feature, which makes your model directly usable in the Transformers library
+    - or leverage the [`huggingface_hub`'s `PyTorchModelHubMixin` class](#upload-a-pytorch-model-using-huggingface_hub) as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module` class, just like models in the Transformers, Diffusers and Timm libraries.
 - In addition to programmatic uploads, you can always use the [web interface](#using-the-web-interface).
 
-Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model.
+Do note that only the first 2 ways will make sure that download metrics work for your model.
```
Comment (osanseviero) [inline docs/hub/models-uploading.md:15]:
```suggestion
Do note that only the first two ways will ensure download metrics work for your model.
```

Comment (osanseviero) [reply docs/hub/models-uploading.md:15]: Maybe add a link to https://huggingface.co/docs/hub/models-download-stats

**Review by osanseviero — COMMENTED** [review]
Very nice :fire:

File: docs/hub/models-uploading.md (hunk 5 above)
Comment (NielsRogge) [reply docs/hub/models-uploading.md:12]: Great tip, added it!

**Review by pcuenca — COMMENTED** [review]
Very cool!

**Comment by NielsRogge** [comment]
Thanks for your reviews, feel free to approve :)

**Review by Wauplin — APPROVED** [review]
Thanks for iterating on that doc!

File: docs/hub/models-uploading.md [hunk 7]
```diff
@@ -4,13 +4,17 @@ To upload models to the Hub, you'll need to create an account at [Hugging Face](
 
 You can link repositories with an individual user, such as [osanseviero/fashion_brands_patterns](https://huggingface.co/osanseviero/fashion_brands_patterns), or with an organization, such as [facebook/bart-large-xsum](https://huggingface.co/facebook/bart-large-xsum). Organizations can collect models related to a company, community, or library! If you choose an organization, the model will be featured on the organization’s page, and every member of the organization will have the ability to contribute to the repository. You can create a new organization [here](https://huggingface.co/organizations/new).
 
-There are several ways to upload models to the Hub, described below.
+> **_NOTE:_** Models do NOT need to be compatible with the Transformers/Diffusers libraries to get download metrics. Any custom PyTorch model is supported. Read more below!
```
Comment (osanseviero) [inline docs/hub/models-uploading.md:7]: This read too pytorch-centric. People can use sklearn/TF/etc

```suggestion
> **_NOTE:_** Models do NOT need to be compatible with the Transformers/Diffusers libraries to get download metrics. Any custom model is supported. Read more below!
```

File: docs/hub/models-uploading.md [hunk 8]
```diff
@@ -4,13 +4,17 @@ To upload models to the Hub, you'll need to create an account at [Hugging Face](
...
+There are several ways to upload models to the Hub and get download metrics, described below.
+
+- In case your model is designed for a library that has [built-in support](#upload-from-a-library-with-built-in-support), you can use the methods provided by the library. Custom models that use `trust_remote_code=True` can also leverage these methods.
+- In case your model is a custom PyTorch model, one can leverage the [`PyTorchModelHubMixin` class](#upload-a-pytorch-model-using-huggingface_hub) as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to any `nn.Module` class, just like models in the Transformers, Diffusers and Timm libraries.
 - In addition to programmatic uploads, you can always use the [web interface](#using-the-web-interface).
 
-Once your model is uploaded, we suggest adding a [Model Card](./model-cards) to your repo to document your model.
+Do note that only the first two ways will ensure [download metrics](models-download-stats) work for your model.
```
Comment (osanseviero) [inline docs/hub/models-uploading.md:15]: Although people can also do their own files and update model-libraries.ts, as https://github.com/huggingface/huggingface.js/pull/561/files

File: docs/hub/models-uploading.md [hunk 9]
```diff
@@ -4,13 +4,17 @@ To upload models to the Hub, you'll need to create an account at [Hugging Face](
...
 
-There are several ways to upload models to the Hub, described below.
+> **_NOTE:_** Models do NOT need to be compatible with the Transformers/Diffusers libraries to get download metrics. Any custom PyTorch model is supported. Read more below!
 
-- In case your model comes from a library that has [built-in support](#upload-from-a-library-with-built-in-support), one can use the existing methods.
-- In case your model is a custom PyTorch model, the recommended way is to leverage the [huggingface_hub](#using-the-huggingface_hub-client-library) Python library as it allows to add `from_pretrained`, `push_to_hub` and [automated download metrics](https://huggingface.co/docs/hub/models-download-stats) capabilities to your models, just like models in the Transformers, Diffusers and Timm libraries.
+There are several ways to upload models to the Hub and get download metrics, described below.
```
Comment (julien-c) [inline docs/hub/models-uploading.md:9]:
```suggestion
There are several ways to upload models for them to be nicely integrated into the Hub and get download metrics, described below.
```

File: docs/hub/models-uploading.md (hunk 8 above)
Comment (julien-c) [reply docs/hub/models-uploading.md:15]: i would just remove the focus on download numbers from this doc page, and just link to the [relevant doc page](https://huggingface.co/docs/hub/models-download-stats)

We can still mention the PytorchMixin

Comment (julien-c) [inline docs/hub/models-uploading.md:15]:
```suggestion
```

**Review by julien-c — COMMENTED** [review]
lgtm with the suggestions
